pkg/groupsync/groups.go

```diff
@@ -0,0 +1,187 @@
+
+	// GetGroup retrieves the Group with the given ID.
+	GetGroup(ctx context.Context, groupID string) (*Group, error)
+
+	// GetMembers retrieves the direct members (children) of the group with given ID.
+	GetMembers(ctx context.Context, groupID string) ([]Member, error)
+
+	// GetUser retrieves the User with the given ID.
+	GetUser(ctx context.Context, userID string) (*User, error)
+}
+
+// GroupWriter provides write operations for a group system.
+type GroupWriter interface {
+	// SetMembers replaces the members of the group with the given ID with the given members.
+	SetMembers(ctx context.Context, groupID string, members []Member) error
+}
+
+// GroupReadWriter provides both read and write operations for a group system.
+type GroupReadWriter interface {
+	GroupReader
+	GroupWriter
+}
+
+// OneToManyGroupMapper maps group IDs to lists of group IDs.
+type OneToManyGroupMapper interface {
+	// AllGroupIDs returns the set of groupIDs being mapped (the key set).
+	AllGroupIDs(ctx context.Context) ([]string, error)
+
+	// ContainsGroupID returns whether this mapper contains a mapping for the given group ID.
+	ContainsGroupID(ctx context.Context, groupID string) (bool, error)
+
+	// MappedGroupIDs returns the list of group IDs mapped to the given group ID.
+	MappedGroupIDs(ctx context.Context, groupID string) ([]string, error)
+}
+
+// UserMapper maps a user ID to another user ID.
+type UserMapper interface {
+	// MappedUserID returns the user ID mapped to the given user ID.
+	MappedUserID(ctx context.Context, userID string) (string, error)
+}
+
+// User represents a user in a group system.
+type User struct {
+	// ID is the user's ID in the group system.
+	ID string `json:"id,omitempty"`
+	// Attributes represent arbitrary attributes about the user
+	// in the given group system. This field is typically set by
+	// the corresponding GroupReader when retrieving the user.
+	Attributes any `json:"attributes,omitempty"`
+}
+
+// Group represents a group in a group system.
+type Group struct {
+	// ID is the group's ID in the group system.
+	ID string `json:"id,omitempty"`
+	// Attributes represent arbitrary attributes about the group
+	// in the given group system. This field is typically set by
+	// the corresponding GroupReader when retrieving the group.
+	Attributes any `json:"attributes,omitempty"`
+}
+
+// Member represents a member of a group. A member may either be
+// a User or another Group. An instance of Member will always be
+// either a User or a Group but not both.
+type Member interface {
+	// IsGroup returns whether this Member is a Group.
+	IsGroup() bool
+
+	// IsUser returns whether this Member is a User.
+	IsUser() bool
+
+	// Group returns the underlying group if this Member is a group.
+	// Otherwise, it returns an error.
+	Group() (*Group, error)
+
+	// User returns the underlying user if this Member is a user.
+	// Otherwise, it returns an error.
+	User() (*User, error)
```

Comment:
nit: Make it clearer that it's _guaranteed_ where will be no error if it's the right member type.